theRAPTLab/meme
# Runtime folder structure and default/Resource packaging updates

This pull request implements the changes requested in #84, namely:

- Data and Resource artifacts separated from the codebase folder and sub-folders to remove the need to re-build, package, and sign when Data/Resources are changed.
- Path decisions largely centralized to support further tweaking that may be necessary for wider cross-platform distribution.
- Resource sets are now included with their respective Data files; previously there were just _datasets_. Now, a _template_ includes both.
- Resources are now seeded with their respective Data files into the top-level _resources_ folder
- Selection of a Data/Resource set ("_template_") has been adjusted to choose from a prioritized list of template options (environment-specified, `meme`, or `blank`)

### Note:
As part of this update, the `static/dlc` folder is no longer present and has been replaced with a top-level `resources` folder. It is therefore necessary that the Express server remap the `static/dlc` URL path to this folder to maintain backward compatibility with older datafiles.

@benloh has indicated that legacy file support may not be necessary. If so, it would be preferred to map `/` within a datafile to be `/resources/` for the purposes of serving content via URL (e.g. https://..../resources/fishtank.pdf) and its location on disk `/<meme folder>/resources/`.


#  Test Instructions

## Test basic paths

1. `nvm use`
2. `npm run clean:all` -- note the `dist` and `built` folders have been removed
3. `npm ci`
4. `npm run dev` -- compile to make sure it runs
5. Check `http://localhost:3000` and `http://localhost:3000/#/admin` to make sure app loads
6. Make sure you prepped the files using the "Preparing for Electron" instructions above.
7. `npm run package`
8. Check `boilerplate/dist/MEME macOS` to make sure the electron app has been built correctly:
     * `data/` -- should be empty
     * `resources/` -- should be empty 
     * `templates/` -- should have a `_blank` template folder
     * `LICENSE`
     * `LICENSE.chromium.html`
     * `meme.app`
     * `version`
9. If you haven't already set up code signing, see [[Signing](https://github.com/theRAPTLab/meme/blob/dev-src/file-paths/README-signing.md)](https://github.com/theRAPTLab/meme/blob/dev-src/file-paths/README-signing.md)
10. Verify that you can run `meme.app` locally on your machine -- start the Electron app, and load `http://localhost:3000/#/admin`
11. `npm run appsign`



## Test create a blank database with `/templates/_blank`

Try different methods of starting a new project from scratch.  You should always be able to create a new blank project without any errors.

### 1. Test `npm run dev` with blank

1. Delete `<repodir>/data` and `dist`
2. Run `npm run dev`
3. A new `test.loki` file should be created in `<repodir>/data/db/test.loki`
4. Going to `http://localhost:3000/#/admin` you should see a blank project

### 2. Test `npm run electron`  with blank

1. Delete `<repodir>/data` and `dist`
2. Run `npm run electron`
3. A new `meme.loki` file should be created in `<repodir>/data/db/meme.loki`
4. An electron app should be built and started
5. Going to `http://localhost:3000/#/admin` you should see a blank project

### 3. Test `npm run app`  with blank

1. Delete `<repodir>/data` and `dist`

2. Run `npm run app` => npm should complain 

   ```
   ./dist/meme-darwin-x64/meme.app: No such file or directory
   MEME EXEC    - ERROR from codesign check
   MEME EXEC    - macos will not run this app until it is signed
   ```

3. Run `npm run package`

4. Run `npm run app` => npm should complain 

   ```
   ./dist/meme-darwin-x64/meme.app: No such file or directory
   MEME EXEC    - ERROR from codesign check
   MEME EXEC    - macos will not run this app until it is signed
   ```

5. Run `npm run electron` to build the `meme.loki` file.

6. Run `npm run package` to build with the necessary depdencies.

7. Run `npm run appsign` (NOTE you need to have set up code signing -- see  [[Signing](https://github.com/theRAPTLab/meme/blob/dev-src/file-paths/README-signing.md)](https://github.com/theRAPTLab/meme/blob/dev-src/file-paths/README-signing.md))

8. A new `meme.loki` file should be created in `<repodir>/data/db/meme.loki`

9. An electron app should be built and started

10. Going to `http://localhost:3000/#/admin` you should see a blank project



## Test with default `/templates/test` template folder

Make sure that `npm run dev` will generate `test.loki` based on a starting template and that it'll fall back to `templates/_blank` if `templates/test` is not defined.

Test Blank

1. Delete `<repodir>/data` and `dist`
2. Delete `/<repodir>/templates/test` (if it exists)
3. `npm run dev`
4. The app should run with a blank project

Test `test`

1. Delete `<repodir>/data` and `dist`
2. Add `/<repodir>/templates/test` -- use the zip file attached.
3. `npm run dev`
4. The app should run with a test project that has three teachers added.



## Test with default `/templates/meme` template folder

Make sure the various ways of setting a starting non-blank template work.

Test Blank

1. Delete `<repodir>/data` and `dist`
2. Delete `/<repodir>/templates/meme` (if it exists)
3. `npm run dev`
4. The app should run with a blank project

Test `npm run electron` with  `meme`

1. Delete `<repodir>/data` and `dist`
2. Add `/<repodir>/templates/meme` -- use the zip file attached.
3. `npm run electron`
4. The app should run with a test project that has three teachers added.

Test `npm run app` with  `meme`

1. Delete `<repodir>/data` and `dist`
2. Add `/<repodir>/templates/meme` -- use the zip file attached.
3. `npm run electron`
4. `npm run package`
5. `npm run appsign`
6. `npm run app`
7. The app should run with a test project that has three teachers added.

Test `meme.app` with  `meme`

1. Delete `<repodir>/data` and `dist`
2. Add `/<repodir>/templates/meme` -- use the zip file attached.
3. `npm run electron`
4. `npm run package`
5. `npm run appsign`
6. Double-click `meme.app` to start Electron
7. The app should run with a test project that has three teachers added.



## Test copy distribution

Make sure it's possible to copy the distribution files to another computer.

1. Build the electron app (`npm run package`) and copy the `/dist/MEME macOS folder` to your new computer/new location, e.g. to `~/Desktop/MEME macOS`.

2. Open a terminal.
3. `cd ~/Desktop/MEME macOS`
4. `./install.sh`
5. The quarantine flag should now be fixed.
6. Double click the meme.app Electron application to start the server.
7. If you see javascript error, then the quarantine flag was probably not removed successfully. Try typing `xattr meme.app` -- you should NOT see the line `com.apple.quarantine`
   If you copy the app to another computer you may have to run `./install.sh` again.
